rust/parser/mod.rs

```diff
@@ -49,279 +46,381 @@ use crate::{
+    outer.into_children().consume_any()
 }
 
-fn get_date_time(date_time: Rc<TerminalNode>) -> NaiveDateTime {
-    date_time::parse(&date_time.get_text())
-        .expect(&ILLEGAL_GRAMMAR.format(&[date_time.get_text().as_str()]).message)
+pub(crate) fn visit_eof_query(query: &str) -> Result<Query> {
+    visit_query(unwrap_single(parse_single(Rule::eof_query, query)?)).validated()
 }
 
-fn get_var(var: Rc<TerminalNode>) -> UnboundVariable {
-    let name = &var.symbol.get_text();
+pub(crate) fn visit_eof_queries(queries: &str) -> Result<impl Iterator<Item = Result<Query>> + '_> {
+    Ok(TypeQLParser::parse(Rule::eof_queries, queries)?
+        .consume_expected(Rule::eof_queries)
+        .into_children()
+        .filter(|child| matches!(child.as_rule(), Rule::query))
+        .map(|query| visit_query(query).validated()))
+}
 
-    assert!(name.len() > 1);
-    assert!(name.starts_with('$'));
-    let name = &name[1..];
+pub(crate) fn visit_eof_pattern(pattern: &str) -> Result<Pattern> {
+    visit_pattern(
+        parse_single(Rule::eof_pattern, pattern)?.into_children().consume_expected(Rule::pattern),
+    )
+    .validated()
+}
 
-    if name == "_" {
-        UnboundVariable::anonymous()
-    } else {
-        UnboundVariable::named(String::from(name))
-    }
+pub(crate) fn visit_eof_patterns(patterns: &str) -> Result<Vec<Pattern>> {
+    visit_patterns(
+        parse_single(Rule::eof_patterns, patterns)?
+            .into_children()
+            .consume_expected(Rule::eof_patterns),
+    )
+    .into_iter()
+    .map(Validatable::validated)
+    .collect()
 }
 
-fn get_isa_constraint(_isa: Rc<TerminalNode>, ctx: Rc<Type_ContextAll>) -> IsaConstraint {
-    match visit_type(ctx) {
-        Type::Label(label) => IsaConstraint::from(label),
-        Type::Variable(var) => IsaConstraint::from(var),
-    }
+pub(crate) fn visit_eof_definables(definables: &str) -> Result<Vec<Definable>> {
+    visit_definables(parse_single(Rule::eof_definables, definables)?)
+        .into_iter()
+        .map(Validatable::validated)
+        .collect()
 }
 
-fn get_role_player_constraint(ctx: Rc<Role_playerContext>) -> RolePlayerConstraint {
-    let player = get_var(ctx.player().unwrap().VAR_().unwrap());
-    if let Some(type_) = ctx.type_() {
-        match visit_type(type_) {
-            Type::Label(label) => RolePlayerConstraint::from((label, player)),
-            Type::Variable(var) => RolePlayerConstraint::from((var, player)),
-        }
-    } else {
-        RolePlayerConstraint::from(player)
-    }
+pub(crate) fn visit_eof_variable(variable: &str) -> Result<Variable> {
+    visit_pattern_variable(parse_single(Rule::eof_variable, variable)?).validated()
 }
 
-fn get_role_players(ctx: Rc<RelationContext>) -> Vec<RolePlayerConstraint> {
-    (0..).map_while(|i| ctx.role_player(i)).map(get_role_player_constraint).collect()
+pub(crate) fn visit_eof_label(label: &str) -> Result<Label> {
+    // TODO validation
```

Comment:
hanging TODO?

Reply:
This TODO is from before and not in scope of this particular PR.